File: Android/app/src/main/java/com/social/app/chat/FollowManager.kt
package com.social.app.chat

import com.social.app.backend.SupabaseManager
import io.github.jan.supabase.postgrest.from
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Gestiona seguir (tabla `follows`, no requiere aceptación mutua como los
 * socials) — equivalente Kotlin de FollowManager.swift. Antes Android no
 * tenía ningún botón ni implementación de "seguir de vuelta" en absoluto.
 *
 * Aviso de honestidad: misma convención de `payload["actor_id"]` ya
 * documentada en FollowManager.swift.
 */
class FollowManager {

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()

    @Serializable
    private data class NewFollow(
        @SerialName("follower_id") val followerId: String,
        @SerialName("followee_id") val followeeId: String
    )

    suspend fun follow(followerId: String, followeeId: String) {
        try {
            SupabaseManager.client.from("follows").insert(NewFollow(followerId, followeeId))
            // Hallazgo real, misma auditoría de AnalyticsManager de la
            // pasada anterior: seguir a alguien tampoco se registraba.
            com.social.app.backend.AnalyticsManager.track("user_followed")
        } catch (e: Exception) {
            _errorMessage.value = "No se pudo seguir a este perfil."
        }
    }

    /**
     * Hallazgo real: no existía ninguna forma de dejar de seguir en ninguna
     * plataforma, ni ningún botón "Seguir" directo en el visor de perfil
     * (solo "seguir de vuelta" desde una notificación) — `follows_select`
     * (0002_rls.sql) es pública (`using (true)`), así que se puede
     * comprobar el estado real sin necesidad de una función RPC.
     */
    suspend fun unfollow(followerId: String, followeeId: String) {
        try {
            SupabaseManager.client.from("follows").delete {
                filter {
                    eq("follower_id", followerId)
                    eq("followee_id", followeeId)
                }
            }
            // Hallazgo real: se registraba `user_followed` pero no su
            // opuesto — sin `user_unfollowed` no se puede medir cuánta
            // gente sigue y luego se arrepiente.
            com.social.app.backend.AnalyticsManager.track("user_unfollowed")
        } catch (e: Exception) {
            _errorMessage.value = "No se pudo dejar de seguir."
        }
    }

    suspend fun isFollowing(followerId: String, followeeId: String): Boolean {
        return try {
            SupabaseManager.client.from("follows")
                .select {
                    filter {
                        eq("follower_id", followerId)
                        eq("followee_id", followeeId)
                    }
                }
                .decodeList<NewFollow>()
                .isNotEmpty()
        } catch (e: Exception) {
            false
        }
    }
}

File: Social/Chat/FollowManager.swift
//
//  FollowManager.swift
//  Social
//
//  Gestiona seguir/dejar de seguir (tabla `follows`, no requiere aceptación
//  mutua como los socials — cualquiera puede seguir a un perfil público).
//  Antes el botón "Seguir de vuelta" en AvisosView.swift estaba vacío
//  (`{}`), sin ninguna implementación.
//
//  Aviso de honestidad: asume que el backend rellena `payload["actor_id"]`
//  con el profile_id de quien generó la notificación de tipo "follow" —
//  misma convención ya usada (y ya documentada como asunción) para
//  `payload["chat_id"]`/`payload["social_id"]` en notificaciones de tipo
//  "social". Si esa pieza server-side no rellena esa clave, el botón no
//  tiene a quién seguir; no es un bug del cliente.
//

import Foundation

@MainActor
final class FollowManager: ObservableObject {

    @Published var errorMessage: String?

    func follow(followerID: UUID, followeeID: UUID) async {
        struct NewFollow: Encodable {
            let follower_id: UUID
            let followee_id: UUID
        }
        do {
            try await SupabaseManager.shared.client
                .from("follows")
                .insert(NewFollow(follower_id: followerID, followee_id: followeeID))
                .execute()
            // Hallazgo real, mismo criterio ya aplicado en la versión
            // Kotlin equivalente: seguir a alguien tampoco se registraba.
            AnalyticsManager.track("user_followed")
        } catch {
            errorMessage = "No se pudo seguir a este perfil."
        }
    }

    /// Hallazgo real: no existía ninguna forma de dejar de seguir en
    /// ninguna plataforma, ni ningún botón "Seguir" directo en el visor de
    /// perfil (solo "seguir de vuelta" desde una notificación) —
    /// `follows_select` (0002_rls.sql) es pública (`using (true)`), así que
    /// se puede comprobar el estado real sin necesidad de una función RPC.
    func unfollow(followerID: UUID, followeeID: UUID) async {
        do {
            try await SupabaseManager.shared.client
                .from("follows")
                .delete()
                .eq("follower_id", value: followerID)
                .eq("followee_id", value: followeeID)
                .execute()
            // Hallazgo real, mismo criterio ya aplicado en la versión
            // Kotlin equivalente: se registraba `user_followed` pero no
            // su opuesto.
            AnalyticsManager.track("user_unfollowed")
        } catch {
            errorMessage = "No se pudo dejar de seguir."
        }
    }

    func isFollowing(followerID: UUID, followeeID: UUID) async -> Bool {
        struct FollowRow: Decodable {
            let follower_id: UUID
            let followee_id: UUID
        }
        let rows: [FollowRow]? = try? await SupabaseManager.shared.client
            .from("follows")
            .select()
            .eq("follower_id", value: followerID)
            .eq("followee_id", value: followeeID)
            .execute()
            .value
        return !(rows ?? []).isEmpty
    }
}
